### DISCUSSION TITLE
Backup files

### DISCUSSION THREAD

**Discussion by 00sapo** [helix-editor/helix#7618] [Ideas]
Basically, the same as the backup option in (n)vim: every time a file is changed and then saved to disk, just save its previous version somewhere to keep a backup of the last edit.

This should be easy to implement, as it is basically a modified version of `:update`. With the variable expansion feature, this could also be set up by users autonomously, but I think it's better to provide a command ad-hoc for this, with a configuration `automatic_backup` and another `backup_directory`. If `backup_directory` is `"."`, then files are backed up side-to-side to the original files with the addition of `~` to the filename.

This is super-useful when while developing you delete files by mistake.

**Comment by EriKWDev** [comment]
When my battary died on windows and later bootet up, the file I had opened in helix was just blank when I opened it again.

I would love an option to just keep rolling backups of the open buffers as well somewhere. Or even auto save after a certain time :(

**Reply by kirawi** [reply]
https://github.com/helix-editor/helix/pull/9236

**Reply by kirawi** [reply]
Ah, actually it's https://github.com/helix-editor/helix/issues/3451

**Reply by EriKWDev** [reply]
While I appreciate the auto save and might try that, given that I had saved the buffer several times but the file still ended up blank, it would really help with the peace of mind of knowing that there could be a rolling backup somewhere else in a hissen file in my home directory or anywhere really

**Reply by kirawi** [reply]
It's strange that happened. Helix shouldn't have been writing while the computer was on standby mode. Unless you meant your computer lost power while saving in which case the original PR would fix that. Can you share a little more about the scenario?

**Reply by EriKWDev** [reply]
It has happened before sadly. When having a file open in helix that I've worked on for a while, regularly saving as usual, and then my laptop's battery dies. When I charge it and come back the file is just completely empty

Has happened on both Windows and Linux. Don't know if it's helix specific or if any editor would do that - I only use helix

**Reply by kirawi** [reply]
If it happens while saving, essentially after `:w`, then https://github.com/helix-editor/helix/pull/9236 would indeed fix it. There are for sure some things we can improve in generally to achieve parity with vim: https://neovim.io/doc/user/recover.html

**Reply by EriKWDev** [reply]
I'm once again asking for your support :( It happened again

My computer crashed due to a GPU memory leak in an application I'm writing. Once I restarted and opened the file I had open in helix, the file was just empty. The file had been saved regularly several times while working.

This happens both on my main desktop and my laptop.

Computer crash while helix open = high likely hood file is gone

```yml
helix 24.3 (00e9e5ea)
```

**Reply by kirawi** [reply]
The fix is on master, not in the release. You can try to build from source or wait until the next release

**Reply by EriKWDev** [reply]
> The fix is on master, not in the release. You can try to build from source or wait until the next release

It is master, see commit version in my message

**Reply by kirawi** [reply]
Ah, my bad. That is very peculiar, Helix should be saving it first to `file_name + .bck` which is then moved to the file path of the original file. So it _shouldn't_ be able to truncate the entire file. It could be something else, like the formatter sending an empty change that truncates the entire file. What language are you using? 

Unfortunately, this is hard to diagnose without being able to reproduce it reliably. I'll look into writing an integration test that would 'crash' Helix mid-write to check if the system is working as expected, but that might take at least 2 weeks since I'm busy right now.

**Reply by kirawi** [reply]
Ah derp, I got it mixed up. The original file should be in `file_name + .bck`. Can you check?

**Reply by the-mikedavis** [reply]
I wonder if we need to fsync(2) the file after writing here? https://github.com/helix-editor/helix/blob/00e9e5eadef16dd20cd24d303a664faaeb8faa56/helix-view/src/document.rs#L931-L933

I'm not sure that `<tokio::fs::File as tokio::io::AsyncWriteExt>::flush` does file syncing and if we don't explicitly sync then the kernel may hold the changes in memory from what I understand. (Although fsync(2) is not always respected by the underlying storage anyways)

**Reply by EriKWDev** [reply]
I want to make a thing clear, the computer doesn't crash in the middle of a save/":w", the computer just crashes for different reasons and helix happens to have the file open. 

This is when, later when I've rebooted, the file is suspiciously completely empty.

I don't know if this happens if I were to use other editors since I use nothing but helix, but it has happened so many times now.. wether it is just battery going out while helix is open or complete computer crash.

If you have any idea how I can help I will happily do testing

**Reply by the-mikedavis** [reply]
It's a bit of a shot in the dark but you can try the syncing suggestion with a small change locally, https://github.com/the-mikedavis/helix/commit/9eb1e594a1d4892625e3282dd3b5f3443855267e

fsync(2) is meant to flush the changes to a file to disk and it's typically used by databases (for example against a write-ahead-log file) to try to ensure that a write isn't lost during a power outage / hard shutdown. Without syncing a write might not actually update the disk, so even if the machine is crashing some time after a `:w`, the changes to the file might still only be in memory and could be lost.

**Reply by the-mikedavis** [reply]
Actually the Tokio docs talk about `flush` not synchronizing. I've posted https://github.com/helix-editor/helix/pull/10735 with that patch to fix this (if that is the issue here)

**Reply by EriKWDev** [reply]
I'm very glad you're looking into this and if the fsync fix is what's needed I will definately give it a try going forward!

**Comment by oderwat** [comment]
I would like that it normally just saves regularly into a backup file while being open for desaster discovery. This could be versioned too. And I need he possibility to enable per language or better per project to save to the real files. This is because I sometimes want to have file watchers trigger rebuilds and hotloading and on other times I don't but still want to be able to recover my work. Having a watcher checking for changes in the buffer would be nice too. It should tell me if there are external changes, long before I would manual safe, because this could invalidate all the work I have done.

P.S.: As a workaround I currently snapshot my workspace using Kopia every 10 minutes.

**Reply by kirawi** [reply]
This is already addressed in https://github.com/helix-editor/helix/pull/11374 (not yet merged) with `editor.backup-kind = copy`

For a file watcher see https://github.com/helix-editor/helix/issues/1125

**Reply by oderwat** [reply]
I don't see that this is doing regular backups while working. It also seems not to have feature, that you for example get an initial copy and the have autosaves or that you get versioned copies while working. Something like this is done by the Jetbrains IDEs for example by implementing the file histories.

**Reply by kirawi** [reply]
Versioning should be the job of a VCS IMO. But https://github.com/helix-editor/helix/pull/9154 will implement persistent sessions once I get around to finishing it though I guess it's not really what you're looking for? I'd suspect the maintainers would prefer such a feature of retaining copies of a file be delegated to plugins. I'm not aware of any other mainstream terminal editor that has such a feature built-in either.

**Reply by oderwat** [reply]
you do not want to store another version ever time you are idle with editing and saving a new version to the backup. I guess my "idea" about what could be done is not understood well. It is not a "version" it is like a persistend undo history, but simplified by storing diffs after editing parts of the text. You then can browser those. Jetbrains for example also stores checkpoints when you run formatters or external programms and also when it reloads changes like adding imports in the head of the files when auto imports are generated. I know it can works as they do that with all files in our huge projects. It is very powerful to have this even if you seldom use it.
